Framework/API/inc/MantidAPI/DetectorInfo.h

```diff
@@ -0,0 +1,136 @@
+  such as Poco threads or Intel TBB.
+
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_API_DLL DetectorInfo {
+public:
+  DetectorInfo(boost::shared_ptr<const Geometry::Instrument> instrument,
+               Geometry::ParameterMap *pmap = nullptr);
+
+  bool isMonitor(const size_t index) const;
+  bool isMasked(const size_t index) const;
+  double l2(const size_t index) const;
+  double twoTheta(const size_t index) const;
+  double signedTwoTheta(const size_t index) const;
+  Kernel::V3D position(const size_t index) const;
+  Kernel::Quat rotation(const size_t index) const;
+
+  void setPosition(const size_t index, const Kernel::V3D &position);
+  void setRotation(const size_t index, const Kernel::Quat &rotation);
+
+  void setPosition(const Geometry::IComponent &comp, const Kernel::V3D &pos);
+  void setRotation(const Geometry::IComponent &comp, const Kernel::Quat &rot);
+
+  // This does not really belong into DetectorInfo, but it seems to be useful
+  // while Instrument-2.0 does not exist.
+  Kernel::V3D sourcePosition() const;
+  Kernel::V3D samplePosition() const;
+  double l1() const;
+
+  const std::vector<detid_t> detectorIDs() const;
+  /// Returns the index of the detector with the given detector ID.
+  size_t indexOf(const detid_t id) const { return m_detIDToIndex.at(id); }
+
+  friend class SpectrumInfo;
+
+private:
+  const Geometry::IDetector &getDetector(const size_t index) const;
+  boost::shared_ptr<const Geometry::IDetector>
+  getDetectorPtr(const size_t index) const;
+  void setCachedDetector(
+      size_t index,
+      boost::shared_ptr<const Geometry::IDetector> detector) const;
+  const Geometry::IComponent &getSource() const;
+  const Geometry::IComponent &getSample() const;
+
+  void cacheSource() const;
+  void cacheSample() const;
+
+  // These cache init functions are not thread-safe! Use only in combination
+  // with std::call_once!
+  void doCacheSource() const;
+  void doCacheSample() const;
+  void cacheL1() const;
+
+  Geometry::ParameterMap *m_pmap;
```

Comment:
As long as we have parametrized instruments and parametrized detectors there will almost always be multiple owners of the `ParameterMap`. Once we managed to removed those we can think about who the actual owner should be. Right now I do not think that it should be `DetectorInfo`, since the parameter will actually mostly contain information about components that are *not* detectors.